File: packages/python/minions/storage/filter_utils.py
"""
minions.storage.filter_utils
=============================
Shared filtering and sorting logic used by all storage adapters.
"""

from __future__ import annotations

from typing import Optional

from ..types import Minion
from .adapter import StorageFilter


def apply_filter(minions: list[Minion], filter: StorageFilter) -> list[Minion]:
    """
    Apply a :class:`StorageFilter` to a list of minions.

    Handles soft-delete exclusion, field-level filtering (type, status, tags),
    sorting, and pagination (limit / offset).
    """
    result = minions

    if not filter.include_deleted:
        result = [m for m in result if not m.deleted_at]
    if filter.minion_type_id is not None:
        result = [m for m in result if m.minion_type_id == filter.minion_type_id]
    if filter.status is not None:
        result = [m for m in result if m.status == filter.status]
    if filter.tags:
        result = [m for m in result if all(t in (m.tags or []) for t in filter.tags)]

    # ── Sorting ──────────────────────────────────────────────────────────────
    if filter.sort_by:
        reverse = filter.sort_order == "desc"

        def _sort_key(m: Minion) -> str:
            if filter.sort_by == "title":
                return m.title.lower()
            elif filter.sort_by == "createdAt":
                return m.created_at
            elif filter.sort_by == "updatedAt":
                return m.updated_at
            return ""

        result = sorted(result, key=_sort_key, reverse=reverse)

    # ── Pagination ───────────────────────────────────────────────────────────
    result = result[filter.offset:]

    if filter.limit is not None:
        result = result[: filter.limit]

    return result

File: packages/core/src/storage/filterUtils.ts
/**
 * @module minions-sdk/storage/filterUtils
 * Shared filtering and sorting logic used by all storage adapters.
 */

import type { Minion } from '../types/index.js';
import type { StorageFilter } from './StorageAdapter.js';

/**
 * Apply a {@link StorageFilter} to an array of minions.
 *
 * Handles soft-delete exclusion, field-level filtering (type, status, tags),
 * sorting, and pagination (limit / offset).
 */
export function applyFilter(minions: Minion[], filter: StorageFilter): Minion[] {
  let result = minions;

  if (!filter.includeDeleted) {
    result = result.filter((m) => !m.deletedAt);
  }
  if (filter.minionTypeId !== undefined) {
    result = result.filter((m) => m.minionTypeId === filter.minionTypeId);
  }
  if (filter.status !== undefined) {
    result = result.filter((m) => m.status === filter.status);
  }
  if (filter.tags && filter.tags.length > 0) {
    result = result.filter((m) =>
      filter.tags!.every((tag) => m.tags?.includes(tag)),
    );
  }

  // ── Sorting ────────────────────────────────────────────────────────────
  if (filter.sortBy) {
    const order = filter.sortOrder === 'desc' ? -1 : 1;
    result = [...result].sort((a, b) => {
      let aVal: string;
      let bVal: string;

      switch (filter.sortBy) {
        case 'title':
          aVal = a.title.toLowerCase();
          bVal = b.title.toLowerCase();
          break;
        case 'createdAt':
          aVal = a.createdAt;
          bVal = b.createdAt;
          break;
        case 'updatedAt':
          aVal = a.updatedAt;
          bVal = b.updatedAt;
          break;
        default:
          return 0;
      }

      return aVal < bVal ? -order : aVal > bVal ? order : 0;
    });
  }

  // ── Pagination ─────────────────────────────────────────────────────────
  const offset = filter.offset ?? 0;
  result = result.slice(offset);

  if (filter.limit !== undefined) {
    result = result.slice(0, filter.limit);
  }

  return result;
}
